### PR TITLE
confd: add the ability to set speed, duplex and autoneg

### PR DESCRIPTION
**Pull request by rical**
Add the ability to configure interface speed, duplex and autoneg. Initially this was pared with YANG model deviations that restricted explicit interface speed and duplex when autoneg was enabled. But due to conflicting default states in YANG and the fact that statd also uses the same YANG meant this code was dropped.

Turning autoneg on (which is the default) is as simple as:
` > set interface x10 ethernet auto-negotiation enable true` 

However, when turning it off the kernel requires an explicit speed and duplex to be set. Ideally we would like to ensure this in the YANG model frontend, as mentioned above. This PR implements a check in the backend witch gives this error when turning off autoneg without an explicit speed and/or duplex (duplex has default):

```
> set interface x10 ethernet auto-negotiation enable false
> leave
Error: Autoneg off requires explicit speed and duplex
Please check your changes, try 'diff' and 'do show interfaces'.
```

The logic here is messy. The fact that `duplex` has a default value and `speed` does not complicates things slightly. The fact that we would prefer not to have an unused `speed` or `duplex` when autoneg is enabled is hard to solve in the backend, as we don't really want to throw an error. Also, the default `duplex` will always exists, making that check impossible without deviating it away(?)

Anyway, I tired to make the code a bit more verbose, with boilerplate like helper functions to make this messy logic easier to follow. Room for improvement here in the future.

### CODE REVIEW

**Review by troglobit — APPROVED** [review]
I think you've made a good trade off here, so for me this gets the nod of approval.  I leave it up to @wkz to weigh in before merging though.

**Comment by wkz** [comment]
I took a stab at some simplifications in e511076. Are there any corner cases I've missed there? Regardless, I think there are some nuggets in there that might be useful.

**Review by wkz — CHANGES_REQUESTED** [review]
I took a stab at some simplifications in https://github.com/kernelkit/infix/commit/e51107606040ef6ab5986624752bc282d293efc4. Are there any corner cases I've missed there? Regardless, I think there are some nuggets in there that might be useful.

**Comment by troglobit** [comment]
e511076 seems to be on a different branch than this PR originates from, but I love the simplification you made, @wkz!  I don't think it's urgent though, so lets await Monday and the return of Mr Alpe for a brief AFK discussion before merging this.

**Comment by rical** [comment]
What a great improvement, thanks a lot @wkz

**Review by wkz — APPROVED** [review]

**Comment by troglobit** [comment]
What's up with the failed jobs?

**Comment by wkz** [comment]
My guess would be that `ethtool -s autoneg on e0` fails when `e0` is a `virtio_net` device.

@rical Can you have a look at this? We may need some prologue in the generated ethtool script that exits early if the interface does not support ethtool. I think `ethtool -i <IFACE>` may be a good way to test support.

**Comment by troglobit** [comment]
Would it help to use the e1000 emulation in Qemu instead?

**Comment by rical** [comment]
> Would it help to use the e1000 emulation in Qemu instead?
It does, but I think we would like to support virtio as well. Discussions ongoing.

**Comment by rical** [comment]
This PR is blocked for now.

We need a new interface type "etherlike" to avoid trying to configure autoneg for virtio devices and such. I'm planning on picking that up next week.

**Comment by troglobit** [comment]
Here we go, amazing work and a huge load of patience went into this one. Thank you, @rical! :astonished: :clinking_glasses: :partying_face:

**Pull request merged**
